### PR TITLE
Reload on config change

### PR DESCRIPTION
**Pull request by tvincent2**
Hello!

Here is finally my attempt at closing #178 . Let me know if anything needs to be changed.

Cheers,
Thomas

### CODE REVIEW

**Comment by Keats** [comment]
Can you get rid of the first 3 commits? I'll have a look at the code itself a bit later today

File: src/cmd/serve.rs [hunk 1]
```diff
@@ -196,6 +199,11 @@ pub fn serve(interface: &str, port: &str, output_dir: &str, config_file: &str) -
                                 console::info(&format!("-> Sass file changed {}", path.display()));
                                 rebuild_done_handling(&broadcaster, site.compile_sass(&site.base_path), &p);
                             },
+                            (ChangeKind::Config, _) => {
+                                console::info(&format!("-> Config changed {}", path.display()));
```
Comment (Keats) [inline src/cmd/serve.rs]: there is only one config file, no need to show the path. The message should say that the whole site will be reloaded and will need to refresh the browser

File: src/cmd/serve.rs [hunk 2]
```diff
@@ -196,6 +199,11 @@ pub fn serve(interface: &str, port: &str, output_dir: &str, config_file: &str) -
...
+                                let path = env::current_dir().unwrap();
```
Comment (Keats) [inline src/cmd/serve.rs]: no need for that line to be here, it can be in `after_config_change` directly

File: src/cmd/serve.rs [hunk 3]
```diff
@@ -249,6 +257,8 @@ fn detect_change_kind(pwd: &str, path: &Path) -> (ChangeKind, String) {
         ChangeKind::StaticFiles
     } else if path_str.starts_with("/sass") {
         ChangeKind::Sass
+    } else if path_str.starts_with("/config.toml") {
```
Comment (Keats) [inline src/cmd/serve.rs]: we can do `path_str == "/config.toml"` directly rather than `starts_with`

File: src/rebuild.rs [hunk 4]
```diff
@@ -254,3 +254,21 @@ pub fn after_template_change(site: &mut Site, path: &Path) -> Result<()> {
         },
     }
 }
+
+/// What happens when the config is changed
+pub fn after_config_change(path: &Path, interface: &str, port: &str, config_file: &str) -> Result<Site> {
```
Comment (Keats) [inline src/rebuild.rs]: this looks suspiciously like the the beginning of the `serve` function, can this just be extracted from it and stay in `serve.rs`?

**Review by Keats — CHANGES_REQUESTED** [review]

File: src/cmd/serve.rs [hunk 5]
```diff
@@ -196,6 +205,10 @@ pub fn serve(interface: &str, port: &str, output_dir: &str, config_file: &str) -
                                 console::info(&format!("-> Sass file changed {}", path.display()));
                                 rebuild_done_handling(&broadcaster, site.compile_sass(&site.base_path), &p);
                             },
+                            (ChangeKind::Config, _) => {
+                                console::info(&format!("-> Config changed. The whole site will be reloaded. The browser needs to be refreshed to make the changes visible."));
+                                site = rebuild::after_config_change(interface, port, output_dir, config_file).unwrap();
```
Comment (Keats) [inline src/cmd/serve.rs]: i think you can just call `create_new_site` here without having anything in `rebuild.rs`

Comment (tvincent2) [reply src/cmd/serve.rs]: Indeed!

**Comment by Keats** [comment]
thanks!

**Pull request merged**